server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -569,21 +601,58 @@ private void createShard(DiscoveryNodes nodes, RoutingTable routingTable, ShardR
                 failedShardHandler,
                 this::updateGlobalCheckpointForShard,
                 retentionLeaseSyncer,
-                nodes.getLocalNode(),
+                originalState.nodes().getLocalNode(),
                 sourceNode
             );
-        } catch (Exception e) {
-            failAndRemoveShard(shardRouting, true, "failed to create shard", e, state);
+            listener.onResponse(true);
+        } catch (ShardLockObtainFailedException e) {
+            logger.warn("shard lock currently unavailable for [{}], retrying in [{}]", shardRouting, shardLockRetryInterval);
```

Comment:
Not easily as things stand because we stop retrying (and start a new retry loop) on each cluster state update. But possibly, I think to make the wait bounded we'll need to know this information.